ledger/testchain-generator/src/main.rs

```diff
@@ -0,0 +1,119 @@
+struct Args {
+    /// The number of validators active on the chain.
+    /// (also corresponds to the number of certificates per round)
+    num_validators: usize,
+    /// The number of blocks to generator.
+    num_blocks: usize,
+    /// Load the genesis block for the chain from the specified path instead of generating it.
+    #[clap(long)]
+    genesis_path: Option<String>,
+    /// Set a custom storage path for the ledger.
+    /// By default, it will use the ledger of the first devnet validator.
+    #[clap(long, short = 'p')]
+    storage_path: Option<String>,
+    /// Remove existing ledger if it already exists.
+    #[clap(long, short = 'f')]
+    force: bool,
+    /// The name of the network to generate the chain for.
+    #[clap(long, value_parser=PossibleValuesParser::new(vec![CanaryV0::SHORT_NAME, TestnetV0::SHORT_NAME, MainnetV0::SHORT_NAME]), default_value=TestnetV0::SHORT_NAME)]
+    network: String,
+}
+
+/// Removes an existing ledger (if any) from the filesystem.
+fn remove_ledger(network: u16, storage_mode: &StorageMode, force: bool) -> Result<()> {
+    let path = aleo_std::aleo_ledger_dir(network, storage_mode);
+
+    if path.exists() && force {
+        std::fs::remove_dir_all(&path).with_context(|| "Failed to remove existing ledger")?;
+
+        println!("Removed existing ledger data at {path:?}");
+    } else if path.exists() {
+        bail!("There is already a ledger at {path:?}. Re-run with `--force` if you want to overwrite it");
+    }
+
+    Ok(())
+}
+
+fn main() -> Result<()> {
+    // Enable logging.
+    tracing_subscriber::fmt::init();
+
+    let args = Args::parse();
+
+    match args.network.as_str() {
+        CanaryV0::SHORT_NAME => generate_testchain::<CanaryV0>(args),
+        TestnetV0::SHORT_NAME => generate_testchain::<TestnetV0>(args),
+        MainnetV0::SHORT_NAME => generate_testchain::<MainnetV0>(args),
+        // This is caught by clap.
+        _ => unreachable!(),
+    }
+}
+
+fn generate_testchain<N: Network>(args: Args) -> Result<()> {
+    let mut rng = TestRng::default();
+    let storage_mode =
+        if let Some(path) = args.storage_path { StorageMode::Custom(path.into()) } else { StorageMode::Development(0) };
+
+    remove_ledger(N::ID, &storage_mode, args.force)?;
+
+    let num_validators = args.num_validators;
+    let num_blocks = args.num_blocks;
+
+    println!("Initializing test chain builder with {num_validators} validators");
+    let mut builder: TestChainBuilder<N> = match args.genesis_path {
+        Some(genesis_path) => TestChainBuilder::new_with_quorum_size_and_genesis_block(num_validators, genesis_path),
+        None => TestChainBuilder::new_with_quorum_size(num_validators, &mut rng),
+    }
+    .with_context(|| "Failed to set up test chain builder")?;
+
+    println!("Generating {num_blocks} blocks");
+
+    let mut pos = 0;
+    let mut blocks = vec![];
+
+    while blocks.len() < num_blocks {
+        let batch_size = (num_blocks - blocks.len()).min(100);
```

Comment:
what's the intent behind the batching? block creation isn't that fast, we might as well log after every one of them, possibly including some useful information like the number of transactions - when I used this utility to produce blocks with 2k txs each from an initial pool of 6k, I felt like we might want to later have a param like `txs_per_block` that - when paired with input txs - would populate each block with the desired number of transactions, so more granularity in logging could be useful

Reply:
No worries, it's simple enough to adjust later on if we consider it useful.